### DISCUSSION TITLE
Install service optionally but still install file (wix3.11)

### DISCUSSION THREAD

**Discussion by thefreephysicist** [wixtoolset/issues#7604] [Questions] [answered]
Hi,

I'm working on an installer which has two features. Either/both features require installing the same exe in the same location. I want it such that for one feature it would be installed as a service, the other feature would be installed as just an executable. Regardless of which feature is installed, the file is installed at the same location - I have been trying to create two components (within the same directory - which is probably the problem...):

```
        <Component Id="cmpB8A68336ACB13FCB99813AFBF8C7768A_Service" Guid="<guid_1>">
          <File Id="filC0A81797D07EB7C3ACE940F99AA9FEDC_1" Source="PathTo\SomeExe.exe" KeyPath="yes" />
          <ServiceInstall Id="ServiceId" DisplayName="ServiceDisplayName" Name="ServiceName" Description="Service Desc" Start="auto" Type="ownProcess" Account="LocalSystem" ErrorControl="normal" Vital="yes" Interactive="no" />
          <ServiceControl Id="ServiceId" Name="ServiceName" Start="install" Stop="both" Remove="uninstall" Wait="yes" />
          <ServiceConfig ServiceName="ServiceName" OnInstall="yes" DelayedAutoStart="yes" />
        </Component>
        <Component Id="cmpB8A68336ACB13FCB99813AFBF8C7768A" Guid="<guid_2">
          <File Id="filC0A81797D07EB7C3ACE940F99AA9FEDC_2" Source="PathTo\SomeExe.exe" KeyPath="yes" />
        </Component>
```

And then I have two ComponentGroups

```
<ComponentGroup Id="ExeWithoutService">
...
      <ComponentRef Id="cmpB8A68336ACB13FCB99813AFBF8C7768A" />
...
</ComponentGroup>
```

and 
```
<ComponentGroup Id="ExeWithService">
...
      <ComponentRef Id="cmpB8A68336ACB13FCB99813AFBF8C7768A_Service" />
...
</ComponentGroup>
```

But because two components have the same file in the same directory I get an error message, something along the lines of:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		ICE30: The target file 'qlmjmveu.exe|SomeExe.exe' is installed in '[ProgramFilesFolder]\InstallLoc\' by two different components on an LFN system: 'cmpB8A68336ACB13FCB99813AFBF8C7768A_Service' and 'cmpB8A68336ACB13FCB99813AFBF8C7768A'. This breaks component reference counting.
```

In my Product I have

```
      <Feature Id="MainProgram" Title="The Service Application" Description="The service application" Level="1">
        <ComponentGroupRef Id="ExeWithService" />
      </Feature>
      <Feature Id="ServiceApp" Title="App" Description="The stand alone application" Level="10">
        <ComponentGroupRef Id="ExeWithoutService" />
      </Feature>
```

I don't necessarily like the idea of two components, but I want to install the service if MainApp is installed (or both MainApp and ServiceApp), and without service if only ServiceApp is installed. Any thoughts about how to go about this?

Thanks in advance!

**Comment by BMurri** [comment, marked as answer]
There are ways, some worse than others, but my recommendation would be to recognize the fact that in Windows service binaries, for several reasons (including security concerns) shouldn't be found in any search paths, while binaries that a use will invoke usually should be.

So, I'd place the file in a directory that'll be accessible via a search path for the non-service case, and instead place that exact same file into a different directory (possibly named "service"?) for the service case.
